### PR TITLE
LinearAlgebra: specialize is{hermitian,symmetric} for {Sym,}Tridiagonal

### PR DESCRIPTION
**Pull request by anaveragehuman**

### CODE REVIEW

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 1]
```diff
@@ -170,6 +170,10 @@ adjoint(S::SymTridiagonal) = Adjoint(S)
 Base.copy(S::Adjoint{<:Any,<:SymTridiagonal}) = SymTridiagonal(map(x -> copy.(adjoint.(x)), (S.parent.dv, S.parent.ev))...)
 Base.copy(S::Transpose{<:Any,<:SymTridiagonal}) = SymTridiagonal(map(x -> copy.(transpose.(x)), (S.parent.dv, S.parent.ev))...)
 
+ishermitian(S::SymTridiagonal{<:Real}) = true
+ishermitian(S::SymTridiagonal{<:Complex}) = sum(imag, S.dv) == 0 && sum(imag, S.ev) == 0
```
Comment (SyxP) [inline stdlib/LinearAlgebra/src/tridiag.jl:174]: You require the diagonal to be real, not just the sum of the imaginary parts to be 0. Take `[im 0; 0 -im]` for example.

Comment (AnAverageHuman) [reply stdlib/LinearAlgebra/src/tridiag.jl]: Good catch, I was trying to come up with something faster than `all(x -> imag(x) == 0, S.dv) && all(x -> imag(x) == 0, S.ev)`. Fixed now, and added a few tests.

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 2]
```diff
@@ -170,6 +170,10 @@ adjoint(S::SymTridiagonal) = Adjoint(S)
 Base.copy(S::Adjoint{<:Any,<:SymTridiagonal}) = SymTridiagonal(map(x -> copy.(adjoint.(x)), (S.parent.dv, S.parent.ev))...)
 Base.copy(S::Transpose{<:Any,<:SymTridiagonal}) = SymTridiagonal(map(x -> copy.(transpose.(x)), (S.parent.dv, S.parent.ev))...)
 
+ishermitian(S::SymTridiagonal{<:Real}) = true
+ishermitian(S::SymTridiagonal{<:Complex}) = all(x -> imag(x) == 0, S.dv) && all(x -> imag(x) == 0, S.ev)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/tridiag.jl:174]:
```suggestion
ishermitian(S::SymTridiagonal) = isreal(S.dv) && isreal(S.ev)
```

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 3]
```diff
@@ -608,6 +612,10 @@ transpose(S::Tridiagonal{<:Number}) = Tridiagonal(S.du, S.d, S.dl)
 Base.copy(aS::Adjoint{<:Any,<:Tridiagonal}) = (S = aS.parent; Tridiagonal(map(x -> copy.(adjoint.(x)), (S.du, S.d, S.dl))...))
 Base.copy(tS::Transpose{<:Any,<:Tridiagonal}) = (S = tS.parent; Tridiagonal(map(x -> copy.(transpose.(x)), (S.du, S.d, S.dl))...))
 
+ishermitian(S::Tridiagonal{<:Real}) = S.du == S.dl
+ishermitian(S::Tridiagonal{<:Complex}) = all(x -> imag(x) == 0, S.d) && S.du == adjoint(S.dl)[1,:]
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/tridiag.jl:616]:
```suggestion
ishermitian(S::Tridiagonal) = S.du == adjoint.(S.dl)
```

File: stdlib/LinearAlgebra/test/tridiag.jl [hunk 4]
```diff
@@ -617,4 +617,28 @@ end
...
+    A = Tridiagonal(A.ev, A.dv, A.ev .+ 1)
+    @test !issymmetric(A)
+    @test !ishermitian(A)
+
+    # complex
+    S = SymTridiagonal(randn(5) .+ 1im, randn(4) .+ 1im)
+    @test issymmetric(S)
+    @test !ishermitian(S)
+
+    S = Tridiagonal(S.ev, zero(S.dv), adjoint(S.ev)[1,:])
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/test/tridiag.jl:639]:
```suggestion
    S = Tridiagonal(S.ev, zero(S.dv), adjoint.(S.ev))
```

**Review by dkarrasch — COMMENTED** [review]
Some simplifications.

File: stdlib/LinearAlgebra/src/tridiag.jl (hunk 3 above)
Comment (AnAverageHuman) [reply stdlib/LinearAlgebra/src/tridiag.jl]: I think this is missing `&& isreal(S.d)`.

File: stdlib/LinearAlgebra/src/tridiag.jl (hunk 2 above)
Comment (AnAverageHuman) [reply stdlib/LinearAlgebra/src/tridiag.jl]: Clearer, but slightly slower (by ~50ns) on my machine with `S = SymTridiagonal(rand(100) .+ 0im, rand(100) .+ 0im)`.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/tridiag.jl]: That's what `isreal` is good for, and we use it even internally in `LinearAlgebra` for that purpose.

**Review by dkarrasch — APPROVED** [review]

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 5]
```diff
@@ -170,6 +170,9 @@ adjoint(S::SymTridiagonal) = Adjoint(S)
 Base.copy(S::Adjoint{<:Any,<:SymTridiagonal}) = SymTridiagonal(map(x -> copy.(adjoint.(x)), (S.parent.dv, S.parent.ev))...)
 Base.copy(S::Transpose{<:Any,<:SymTridiagonal}) = SymTridiagonal(map(x -> copy.(transpose.(x)), (S.parent.dv, S.parent.ev))...)
 
+ishermitian(S::SymTridiagonal) = isreal(S.dv) && isreal(S.ev)
```
Comment (sostock) [inline stdlib/LinearAlgebra/src/tridiag.jl:173]: `SymTridiagonal` allows `ev` to have one extra element (i.e., `dv` and `ev` are allowed to have the same length). So, to be correct, should `isreal(S.ev)` be replaced by `isreal(view(S.ev, 1:length(S.dv)-1))`?

**Pull request merged**